### PR TITLE
feat(cdk-graph): initial commit

### PR DESCRIPTION
**Pull request by JeremyJonas**
## CDK Graph (`@aws-prototyping-sdk/cdk-graph`)

experimental

> More comprehensive documentation to come as this package stabilizes

This package is the core framework for supporting additional cdk based automation and tooling, such as diagraming, cost modeling, security and compliance, in a holistic and comprehensive way.

This package provides the following functionality:

1. Synthesizes a serialized graph (nodes and edges) from CDK source code.
1. Provides runtime interface for interacting with the graph (in-memory database-like graph store).
1. Provides plugin framework for additional tooling to utilize and extend the graph.

The goal of this framework is to enable bespoke tooling to be built without having to first traverse the CDK Tree and Metadata to build a graph. Projects like [cdk-dia](https://github.com/pistazie/cdk-dia) generate a bespoke in-memory graph that is then utilized to generate diagrams; while the diagram generation is the core value it must first have a graph to act upon and currently is required to generate this undifferentiated graph to provide its diagrams. By standardizing on the graph interface necessary to build complex tooling, we can more rapidly build new tooling that focuses on its core value.

### CODE REVIEW

File: packages/cdk-graph/README.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+   */
+  report?: IGraphReportCallback;
+}
+```
+
+Plugin operations are automatically invoked by CdkGraph in the order they are defined in the `plugins` property. The invocation flow of plugins follows: 1) `bind`, 2) `inspect`, 3) `synthesize`, 4) `async report`.
+
+### Asynchronous Plugins
+
+Some plugins may requiring performing asynchronous requests, or may make expense operations that are best left outside of the synthesis process.
```
Comment (agdimech) [inline packages/cdk-graph/README.md]: expensive*

File: packages/cdk-graph/README.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    }),
+  ],
+});
+```
+
+### Plugin Interface
+
+```ts
+/** CdkGraph **Plugin** interface */
+export interface ICdkGraphPlugin {
```
Comment (agdimech) [inline packages/cdk-graph/README.md:100]: We should add developer docs on how to implement a custom plugin.

File: packages/cdk-graph/src/cdk-graph.ts [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+
+/**
+ * CdkGraph artifact definition
+ * @struct
+ */
+export interface CdkGraphArtifact {
+  /** The unique type of the artifact */
+  readonly id: string;
+  /** Filename of the artifact */
+  readonly filename: string;
```
Comment (agdimech) [inline packages/cdk-graph/src/cdk-graph.ts:41]: NIT: Should be camelCase.

File: packages/cdk-graph/src/cdk-graph.ts [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+  /** Full path where artifact is stored */
+  readonly filepath: string;
```
Comment (agdimech) [inline packages/cdk-graph/src/cdk-graph.ts:43]: NIT: Should be camelCase.

File: packages/cdk-graph/src/cdk-graph.ts [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+export class CdkGraphContext {
+  /** @internal */
+  readonly _artifacts: { [id: string]: CdkGraphArtifact } = {};
+
+  constructor(
+    public readonly store: Graph.Store,
+    public readonly outdir: string
+  ) {}
+
+  /** Get CdkGraph artifact by id */
```
Comment (agdimech) [inline packages/cdk-graph/src/cdk-graph.ts]: Missing @throws.

File: packages/cdk-graph/src/cdk-graph.ts [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+    const artifact = this._artifacts[id];
+    if (artifact) {
+      return artifact;
+    }
+    throw new Error(`Graph artifact ${id} does not exist`);
+  }
+
+  /** Get CdkGraph core `graph.json` artifact */
+  get graphJson(): CdkGraphArtifact {
+    return this.getArtifact("GRAPH");
```
Comment (agdimech) [inline packages/cdk-graph/src/cdk-graph.ts]: perhaps move "GRAPH" to a literal const.

File: packages/cdk-graph/src/cdk-graph.ts [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+  /** Description of artifact */
+  readonly description?: string;
+  /** The source of the artifact (such as plugin, or core system, etc) */
+  readonly source: string;
+}
+
+/** CdkGraph context */
+export class CdkGraphContext {
+  /** @internal */
+  readonly _artifacts: { [id: string]: CdkGraphArtifact } = {};
```
Comment (agdimech) [inline packages/cdk-graph/src/cdk-graph.ts]: Make `{ [id: string]: CdkGraphArtifact }` a type.

File: packages/cdk-graph/src/cdk-graph.ts [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+
+  /** Indicates if context has an artifact with *filename* defined */
+  hasArtifactFile(filename: string): boolean {
+    return !!Object.values(this._artifacts).find(
+      (artifact) => artifact.filename === filename
+    );
+  }
+
+  /** Get record of all graph artifacts keyed by artifact id */
+  get artifacts(): { [id: string]: CdkGraphArtifact } {
```
Comment (agdimech) [inline packages/cdk-graph/src/cdk-graph.ts]: re-use type here.

File: packages/cdk-graph/src/cdk-graph.ts [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+
+  /**
+   * Logs an artifact entry. In general this should not be called directly, as `writeArtifact` should be utilized
+   * to perform writing and logging artifacts. However some plugins utilize other tools that generate the artifacts,
+   * in which case the plugin would call this method to log the entry.
+   * @param source The source of the artifact, such as the name of plugin
+   * @param id Unique id of the artifact
+   * @param filepath Full path where the artifact is stored
+   * @param description Description of the artifact
+   * @returns
```
Comment (agdimech) [inline packages/cdk-graph/src/cdk-graph.ts:100]: Missing @throws.

File: packages/cdk-graph/src/cdk-graph.ts [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+
+  /** @internal */
+  protected _synthesize(session: ISynthesisSession): void {
+    const store = computeGraph(this.root);
+    const outdir = resolveOutdir(session.outdir, this.config.outdir);
+    const context = new CdkGraphContext(store, outdir);
+
+    context.writeArtifact(
+      this,
+      "GRAPH",
```
Comment (agdimech) [inline packages/cdk-graph/src/cdk-graph.ts]: Reference const

File: packages/cdk-graph/src/cdk-graph.ts [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+      CdkGraphArtifacts.GRAPH,
+      JSON.stringify(context.store.serialize(), null, 2),
+      "Serialized graph"
+    );
+
+    this.plugins.forEach((plugin) => {
+      if (plugin.synthesize) {
```
Comment (agdimech) [inline packages/cdk-graph/src/cdk-graph.ts]: can be simplified to: `plugin.synthesize && plugin.synthesize(context);`

File: packages/cdk-graph/src/cdk-graph.ts [hunk 12]
```diff
@@ -0,0 +1,343 @@
...
+      // TODO: support deserializing pdk-graph to generate store/context
+      console.warn(
+        chalk.yellowBright(
+          "[CdkGraph] In the near future, reports will be runnable outside of cdk synth"
+        )
+      );
+      throw new Error("CdkGraph report called outside of cdk synth process");
+    }
+    for (const plugin of this.plugins) {
+      if (plugin.report) {
```
Comment (agdimech) [inline packages/cdk-graph/src/cdk-graph.ts]: can be simplified to `plugin.report && plugin.report(this._context);`

File: packages/cdk-graph/src/cdk-internals/index.ts [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+ */
+const CUSTOM_SYNTHESIS_SYM = Symbol.for("@aws-cdk/core:customSynthesis");
+
+/**
+ * Interface for constructs that want to do something custom during synthesis
+ *
+ * This feature is intended for use by official AWS CDK libraries only; 3rd party
+ * library authors and CDK users should not use this function.
+ *
+ * **TODO**: Find an alternative synthesis solution that goes not conflict with
```
Comment (agdimech) [inline packages/cdk-graph/src/cdk-internals/index.ts]: goes => does*

File: packages/cdk-graph/src/config/index.ts [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+   * Directory where artifacts are written.
+   *
+   * The key `<cdk.out>` will be replaced with the synthesizer cdk `outdir`.
+   *
+   * Relative paths not prefixed with `<cdk.out>` will be relative to `process.cwd`
+   *
+   * @default "<cdk.out>/cdkgraph"
+   */
+  readonly outdir?: string;
+  readonly [key: string]: any;
```
Comment (agdimech) [inline packages/cdk-graph/src/config/index.ts:40]: missing docs.

File: packages/cdk-graph/src/core/counter.ts [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ ******************************************************************************************************************** */
+/**
+ * Counter class provides utils for adding/subtracting key based counts.
+ *
+ * For example, you can create a counter to track number of each CfnResourceType based on resource type key.
+ */
+export class Counter<T = string> {
```
Comment (agdimech) [inline packages/cdk-graph/src/core/counter.ts:35]: This is only used internally right i.e: not exposed via JSII? 

The reason I ask is that generics I believe are not supported in the JSII spec. Please verify this PR works in the generated Java and Python packages.

File: packages/cdk-graph/src/core/interface.ts [hunk 16]
```diff
@@ -0,0 +1,2644 @@
...
+     */
+    computeLogicalUniversalId(
+      stack: StackNode,
+      logicalId: string
+    ): LOGICAL_UNIVERSAL_ID {
+      return `${stack.uuid}:${logicalId}`;
+    }
+
+    /** Find node by **universal** *logicalId* (`<stack>:<logicalId>`) */
+    findByLogicalUniversalId(uid: LOGICAL_UNIVERSAL_ID): Node {
```
Comment (agdimech) [inline packages/cdk-graph/src/core/interface.ts]: perhaps rename to be more explicit: `findNodeBy...`

File: packages/cdk-graph/src/core/interface.ts [hunk 17]
```diff
@@ -0,0 +1,2644 @@
...
+      const [stackUUID, logicalId] = uid.split(":");
+      const stack: StackNode = this.getStack(stackUUID);
+
+      return this.findByLogicalId(stack, logicalId);
+    }
+
+    /** Find node within given **stack** with given *logicalId* */
+    findByLogicalId(stack: StackNode, logicalId: string): Node {
```
Comment (agdimech) [inline packages/cdk-graph/src/core/interface.ts]: perhaps rename to be more explicit: `findNodeBy...`

File: packages/cdk-graph/src/core/interface.ts [hunk 18]
```diff
@@ -0,0 +1,2644 @@
...
+  /** Edge props interface */
+  export interface IEdgeProps extends ITypedEdgeProps {
+    /** Indicates the direction in which the edge is directed */
+    readonly direction: EdgeDirectionEnum;
+    /** Type of edge */
+    readonly edgeType: EdgeTypeEnum;
+  }
+
+  /**
+   * EdgeCain defines edge traversal path to terminal point.
```
Comment (agdimech) [inline packages/cdk-graph/src/core/interface.ts]: defines an*

File: packages/cdk-graph/src/core/interface.ts [hunk 19]
```diff
@@ -0,0 +1,2644 @@
...
+      this.store.mutateRemoveEdge(this);
+      this._destroyed = true;
+    }
+
+    /**
+     * Merge an equivalent edge's data into this edge and destroy the other edge.
+     *
+     * Used during filtering operations to consolidate equivalent edges.
+     * @param edge - The edge to consume
+     * @destructive
```
Comment (agdimech) [inline packages/cdk-graph/src/core/interface.ts]: missing @throws

File: packages/cdk-graph/src/core/interface.ts [hunk 20]
```diff
@@ -0,0 +1,2644 @@
...
+    mutateHoist(_newParent: Node): void {
+      throw new Error("Root node can not be hoisted");
+    }
+  }
+}
+
+/**
+ * Deserializes a *serialized store object* into an *in-memory store instance**.
+ * @param serializedStore - The serialized store to deserialize
+ * @param allowDestructiveMutations - Indicates if the store instance allows destructive mutations.
```
Comment (agdimech) [inline packages/cdk-graph/src/core/interface.ts]: missing @throws

File: packages/cdk-graph/src/core/interface.ts [hunk 21]
```diff
@@ -0,0 +1,2644 @@
...
+ */
+export function deserializeStore(
+  serializedStore: SerializedGraph.GraphStore,
+  allowDestructiveMutations: boolean = false
+): Graph.Store {
+  const store = new Graph.Store(allowDestructiveMutations);
+
+  // TODO: ensure store versions are compatible
```
Comment (agdimech) [inline packages/cdk-graph/src/core/interface.ts]: Would be good to track all of these in the README.md or gh issue.

File: packages/cdk-graph/src/core/interface.ts [hunk 22]
```diff
@@ -0,0 +1,2644 @@
...
+      }
+      default: {
+        // TODO: support custom edge types explicitly
+        new Graph.Edge(edgeProps);
+      }
+    }
+  });
+
+  return store;
+}
```
Comment (agdimech) [inline packages/cdk-graph/src/core/interface.ts]: I think that is the longest file I have ever reviewed *wipes sweat off forehead*

File: packages/cdk-graph/src/filtering/presets.ts [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+
+/**
+ * Verify that store is filterable, meaning it allows destructive mutations.
+ * @throws Error if store is not filterable
+ * @internal
+ */
+export function verifyFilterable(store: Graph.Store): void {
+  if (!store.allowDestructiveMutations) {
+    throw new Error(
+      "Store must allow destructive mutations to perform filtering; clone store before applying filters"
```
Comment (agdimech) [inline packages/cdk-graph/src/filtering/presets.ts]: be more explicit and give the clone method to call.

File: private/projects/cdk-graph.ts [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+        "@types/lodash.omit",
+        "@types/lodash.uniq",
+        "@types/lodash.uniqby",
+        "@types/traverse",
+        "aws-cdk-lib",
+        "constructs",
+        "projen",
+      ],
+      peerDeps: ["projen", "aws-cdk-lib", "constructs"],
+      bundledDeps: [
```
Comment (agdimech) [inline private/projects/cdk-graph.ts:47]: Why are these all bundled?

**Review by agdimech — COMMENTED** [review]
Great work Jeremy, this is going to be super handy when implementing some exciting plugins :)

File: private/projects/cdk-graph.ts (hunk 24 above)
Comment (JeremyJonas) [reply private/projects/cdk-graph.ts:47]: The implementation requires them, but they are not JSII dependencies, so need to be bundled. At least that is my understanding of how JSII works.

I broken lodash out into explicit methods to reduce the compiled size.

Is there a different way to accomplish this in JSII? @agdimech

File: packages/cdk-graph/README.md (hunk 2 above)
Comment (JeremyJonas) [reply packages/cdk-graph/README.md:100]: Agree - I will add and update the PR

File: packages/cdk-graph/src/cdk-graph.ts (hunk 3 above)
Comment (JeremyJonas) [reply packages/cdk-graph/src/cdk-graph.ts:41]: "filename" is proper english word - https://www.dictionary.com/browse/filename

File: packages/cdk-graph/src/cdk-graph.ts (hunk 4 above)
Comment (JeremyJonas) [reply packages/cdk-graph/src/cdk-graph.ts:43]: While "filepath" is not proper english word - given "filename" is I believe consistency here warrants the above... and maybe in the near future "filepath" like "filename" will become proper :)

But happy to change to "filePath" if disagree

File: packages/cdk-graph/src/cdk-graph.ts (hunk 6 above)
Comment (JeremyJonas) [reply packages/cdk-graph/src/cdk-graph.ts]: Will do

File: packages/cdk-graph/src/cdk-graph.ts (hunk 7 above)
Comment (JeremyJonas) [reply packages/cdk-graph/src/cdk-graph.ts]: Acknowledged

File: packages/cdk-graph/src/cdk-graph.ts (hunk 8 above)
Comment (JeremyJonas) [reply packages/cdk-graph/src/cdk-graph.ts]: Acknowledged

File: packages/cdk-graph/src/core/counter.ts (hunk 15 above)
Comment (JeremyJonas) [reply packages/cdk-graph/src/core/counter.ts:35]: Yes should be, but might ultimately get exposed via the Store getter. I will verify this.

File: packages/cdk-graph/src/core/interface.ts (hunk 16 above)
Comment (JeremyJonas) [reply packages/cdk-graph/src/core/interface.ts]: Acknowledged

File: packages/cdk-graph/src/core/interface.ts (hunk 17 above)
Comment (JeremyJonas) [reply packages/cdk-graph/src/core/interface.ts]: Acknowledged

File: packages/cdk-graph/src/filtering/presets.ts (hunk 23 above)
Comment (JeremyJonas) [reply packages/cdk-graph/src/filtering/presets.ts]: Acknowledged

File: packages/cdk-graph/src/core/interface.ts (hunk 22 above)
Comment (JeremyJonas) [reply packages/cdk-graph/src/core/interface.ts]: I ran into circular dependency issues and JSII issues when splitting this up early on, but I will take another look at how to refactor into smaller modules and post update to PR.

File: packages/cdk-graph/src/core/counter.ts (hunk 15 above)
Comment (JeremyJonas) [reply packages/cdk-graph/src/core/counter.ts:35]: Good catch - https://aws.github.io/jsii/user-guides/lib-author/typescript-restrictions/#parameterized-types-aka-generics

This was intended to be `@internal` but was in fact surfaced from `Store.counts`, I have updated this to JSII compliant record that is exposed and properly defined `@internal` as desired.

File: packages/cdk-graph/src/core/interface.ts (hunk 21 above)
Comment (JeremyJonas) [reply packages/cdk-graph/src/core/interface.ts]: Done, I added a "Path to Stable" section in README with a check list that includes TODO and other areas of coverage.

File: packages/cdk-graph/src/core/interface.ts (hunk 22 above)
Comment (JeremyJonas) [reply packages/cdk-graph/src/core/interface.ts]: As discussed offline, will keep this as is for now and will refactor before stable version.

File: private/projects/cdk-graph.ts (hunk 24 above)
Comment (agdimech) [reply private/projects/cdk-graph.ts:47]: No, it is actually fine how you have it. Ignore my comment.

**Review by agdimech — APPROVED** [review]

**Pull request merged**